shared/dataflow/codeql/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -1314,15 +1482,61 @@ module MakeImplCommon<LocationSig Location, InputSig<Location> Lang> {
     ContentApprox getContentApproxCached(Content c) { result = getContentApprox(c) }
 
     cached
-    newtype TCallContext =
-      TAnyCallContext() or
-      TSpecificCall(DataFlowCall call) {
-        CachedCallContextSensitivity::recordDataFlowCallSite(call, _)
-      } or
-      TSomeCall() or
-      TReturn(DataFlowCallable c, DataFlowCall call) {
-        CachedCallContextSensitivity::reducedViableImplInReturn(c, call)
+    newtype TCallEdge =
+      TMkCallEdge(DataFlowCall call, DataFlowCallable tgt) { viableCallableExt(call) = tgt }
+
+    cached
+    int edgeOrder(TCallEdge edge) {
+      edge =
+        rank[result](TCallEdge e, DataFlowCall call, DataFlowCallable tgt |
+          e = TMkCallEdge(call, tgt)
+        |
+          e order by call.totalorder(), tgt.totalorder()
+        )
+    }
+
+    cached
+    int callOrder(DataFlowCall call) { result = call.totalorder() }
+
+    private module UnreachableSetsInput implements MkSetsInp {
+      class Key = TCallEdge;
+
+      class Value = NodeRegion;
+
+      NodeRegion getAValue(TCallEdge edge) {
+        exists(DataFlowCall call, DataFlowCallable tgt |
+          edge = TMkCallEdge(call, tgt) and
+          getNodeRegionEnclosingCallable(result) = tgt and
+          isUnreachableInCallCached(result, call)
+        )
+      }
+
+      int totalorder(NodeRegion nr) { result = nr.totalOrder() }
+    }
+
+    private module UnreachableSets = MakeSets<UnreachableSetsInput>;
+
+    /** A set of nodes that is unreachable in some call context. */
+    cached
+    class UnreachableSet instanceof UnreachableSets::ValueSet {
+      cached
+      string toString() { result = "Unreachable" }
```

Comment:
I specifically don't want to inherit any member methods, since that would ruin caching. So for caching reasons I strongly prefer `instanceof` over final extends here.